src/treeview_switches.ml

```diff
@@ -0,0 +1,207 @@
+  let collapsibleState = Dependency.collapsible_state dependency in
+  let label =
+    Vscode.TreeItemLabel.create ~label:(Dependency.label dependency) ()
+  in
+  let item = Vscode.TreeItem.make ~label ~collapsibleState () in
+  TreeItem.set_id item (Dependency.to_string dependency);
+  TreeItem.set_iconPath item icon;
+  TreeItem.set_contextValue item (Dependency.context_value dependency);
+  let+ _ =
+    Promise.Option.iter
+      (fun desc -> TreeItem.set_description item (`String desc))
+      (Dependency.description dependency)
+  in
+  Option.iter (Dependency.tooltip dependency) ~f:(fun desc ->
+      TreeItem.set_tooltip item (`String desc));
+  item
+
+let get_dependency_dependencies _dependency = Promise.return (Some [])
+
+let get_switch_dependencies switch =
+  let open Promise.Syntax in
+  let* packages = Opam.get_switch_packages switch in
+  match packages with
+  | Ok packages ->
+    let names =
+      packages
+      |> List.map ~f:(fun n ->
+             Dependency.Dependency (n, Some (Dependency.Switch switch)))
+    in
+    Promise.return (Some names)
+  | Error err ->
+    let* _ =
+      Vscode.Window.showInformationMessage
+        ~message:
+          (Printf.sprintf
+             "An error occured while reading the switch dependencies: %s" err)
+        ()
+    in
+    Promise.return None
+
+let get_dependencies ~extension_path element =
+  let open Promise.Syntax in
+  let dependency =
+    match Vscode.TreeItem.id element with
+    | None ->
+      (* This should never happen, the elements are always created with an ID. *)
+      assert false
+    | Some dep -> Dependency.of_string dep
+  in
+  let dependencies_opt_to_tree_items deps_opt =
+    let* deps_opt = deps_opt in
+    match deps_opt with
+    | None -> Promise.return None
+    | Some deps ->
+      let* items =
+        List.fold_left ~init:(Promise.return [])
+          ~f:(fun acc el ->
+            let* acc = acc in
+            let+ item = make_item ~extension_path el in
+            item :: acc)
+          deps
+      in
+      Promise.return (Some items)
+  in
+  match dependency with
+  | Dependency.Switch switch ->
+    let deps = get_switch_dependencies switch in
+    dependencies_opt_to_tree_items deps
+  | Dependency.Dependency dependency ->
+    let deps = get_dependency_dependencies dependency in
+    dependencies_opt_to_tree_items deps
+
+let register extension =
+  let open Promise.Syntax in
+  let extension_path = Vscode.ExtensionContext.extensionPath extension in
+  let+ opam = Opam.make () in
+  let getChildren ~element =
+    match element with
+    | None ->
+      let items =
+        match opam with
+        | None -> Promise.return None
+        | Some opam ->
+          let* switches = Opam.switch_list opam in
+          let+ items =
+            Promise.List.filter_map
+              (function
+                | Opam.Switch.Named name
+                  when String.is_prefix name ~prefix:"vscode-ocaml-toolchain" ->
```

Comment:
I'd remove this special prefix detection for now, because I'd prefer if we handled private switches differently.

Reply:
Ah yes, I did the private switch modification in the other PR also. Will remove